### ISSUE TITLE
Improvements to the avro processor

### ISSUE THREAD

**Issue by gregorskii** [Jeffail/benthos#381]
HI again,

I am working with the Avro processor. We are trying to use it to process Avro textual records (or pure JSON) as the event input, and marshal it to Avro a record to send to GCP as bytes.

Currently the process we have been debugging in #377 is working, but we are trying to move Avro validation and processing closer to our event processing side (after GCP).

In our ideal flow our event publishers would send either pure JSON or textual Avro records to Benthos, Benthos would process those records into Avro records and transmit them to our pipeline as a Avro byte object. The pipeline would be able to validate the bytes with the Avro reader and writer.

Is this possible with Benthos and the Avro processor currently, or am I misunderstanding what the Avro processor is doing?

I have added it to a test pipeline with a schema defined and it seems to have no affect.

A second question is whether the Avro.schema property can be loaded from a file dynamically, or does it have to come from a single Avro schema definition?

Here is our example Avro schema:

```
{
  "type" : "record",
  "name" : "Event_TestEvent",
  "fields" : [ {
    "name" : "event_type",
    "type" : {
      "type" : "enum",
      "name" : "EventType",
      "symbols" : [ "TestEvent" ]
    }
  }, {
    "name" : "metadata",
    "type" : {
      "type" : "record",
      "name" : "Metadata",
      "fields" : [ {
        "name" : "schema_version",
        "type" : "string"
      } ]
    }
  }, {
    "name" : "event_properties",
    "type" : {
      "type" : "record",
      "name" : "TestEventProperties",
      "fields" : [ {
        "name" : "source",
        "type" : [ "null", "string" ],
        "default" : null
      } ]
    }
  }, {
    "name" : "event_id",
    "type" : "string"
  }, {
    "name" : "timestamp",
    "type" : {
      "type" : "long",
      "logicalType" : "timestamp-millis"
    }
  } ]
}
```

Benthos config:

```
http:
  address: 0.0.0.0:1337
  read_timeout: 5s
  root_path: /benthos
  debug_endpoints: true

input:
  type: broker
  broker:
    copies: 25
    inputs:
    - type: http_server
      http_server:
        address: ""
        cert_file: ""
        key_file: ""
        path: /track
        timeout: 5s
        ws_path: /track/ws

buffer:
  type: memory
  memory:
    limit: 1_000_000_000

pipeline:
  processors:
    - avro:
        operator: from_json
        encoding: textual
        schema: "{\"type\":\"record\",\"name\":\"Event_TestEvent\",\"fields\":[{\"name\":\"event_type\",\"type\":{\"type\":\"enum\",\"name\":\"EventType\",\"symbols\":[\"TestEvent\"]}},{\"name\":\"metadata\",\"type\":{\"type\":\"record\",\"name\":\"Metadata\",\"fields\":[{\"name\":\"schema_version\",\"type\":\"string\"}]}},{\"name\":\"event_properties\",\"type\":{\"type\":\"record\",\"name\":\"TestEventProperties\",\"fields\":[{\"name\":\"source\",\"type\":[\"null\",\"string\"],\"default\":null}]}},{\"name\":\"event_id\",\"type\":\"string\"},{\"name\":\"timestamp\",\"type\":{\"type\":\"long\",\"logicalType\":\"timestamp-millis\"}}]}"

    - log:
        level: INFO
        message: "DATA ${!content}"

output:
  type: stdout

logger:
  prefix: benthos
  level: INFO
  add_timestamp: true
  json_format: true
  static_fields:
    '@service': benthos

shutdown_timeout: 20s
```

Thanks!

**Comment by Jeffail** [comment]
Hey @gregorskii, yeah that should be possible.

The dynamic schema part is harder, do you mean selecting a schema based on a field within each message?

Right now if you want to select from N schemas then you'd need to configure that many processors and use a switch or something to select them, which is pretty gross. It would be quick to add a `schema_file` field as an alternative to pasting the whole schema in the config, and then maybe we can explore allowing function interpolation on that field.

**Comment by gregorskii** [comment]
We can try using multiple schema definitions and a switch to see if it works for now, then enhance it later.

It seemed like the Avro processor had no affect, how do we debug it or test if it was working?

It seemed if we added new fields to the json input the output would have the field even if the schema did not define that field, and the same if we removed them. We tried outputting the content, and checking on the pub sub side to see what the final output would be.

Thanks

**Comment by Jeffail** [comment]
What you should be seeing on the output is a textual avro payload, if you get JSON out then it'll be failing to convert it. Either set logging to debug mode to see if there are messages failing validation or add an error handle to expose it:

```yaml
pipeline:
  processors:
    - avro:
        operator: from_json
        encoding: textual
        schema: "{\"type\":\"record\",\"name\":\"Event_TestEvent\",\"fields\":[{\"name\":\"event_type\",\"type\":{\"type\":\"enum\",\"name\":\"EventType\",\"symbols\":[\"TestEvent\"]}},{\"name\":\"metadata\",\"type\":{\"type\":\"record\",\"name\":\"Metadata\",\"fields\":[{\"name\":\"schema_version\",\"type\":\"string\"}]}},{\"name\":\"event_properties\",\"type\":{\"type\":\"record\",\"name\":\"TestEventProperties\",\"fields\":[{\"name\":\"source\",\"type\":[\"null\",\"string\"],\"default\":null}]}},{\"name\":\"event_id\",\"type\":\"string\"},{\"name\":\"timestamp\",\"type\":{\"type\":\"long\",\"logicalType\":\"timestamp-millis\"}}]}"

    - catch:
      - log:
          message: "Validation failed due to: ${!error}"

    - log:
        level: INFO
        message: "DATA ${!content}"
```

The docs ought to have some usage examples to demo this stuff but I've been waiting to hear back on use cases, this is probably a good case study.

**Comment by gregorskii** [comment]
Yep I see the errors now, looking into it.

**Comment by gregorskii** [comment]
We have gotten our processor to work correctly. We found the issue was related to the `timestamp-millis` field. We set the timestamp field to be configured like this:

```
{
    "name" : "timestamp",
    "type" : {
      "type" : "long",
      "logicalType" : "timestamp-millis"
    }
  }
```

And pass it in JSON as an integer timestamp `"timestamp": 1582064195`. This resulted in a failure due to the GoAvro marshalling rules: https://github.com/linkedin/goavro#translating-from-avro-to-go-data where `int` in JSON is converted to a `float64` in Go. But the Avro processor is falling into this error case:

https://github.com/linkedin/goavro/blob/master/logical_type.go#L133

A fix for this would be to allow the Benthos Avro processor define how date fields are being sent (as float64) and convert them to a known type of avro date. But this would be less dynamic. Is this just a mismatch with how numerical timestamps would be sent in JSON --- which is always an integer?

The exact error we got was this:

```
benthos_1  | {"@timestamp":"2020-02-21T20:53:49Z","@service":"benthos","level":"INFO","component":"benthos.pipeline.processor.1.0","message":"Validation failed due to: failed to convert JSON to Avro schema: cannot encode textual map: value for \"timestamp\" does not match its schema: cannot transform binary timestamp-millis, expected time.Time, received float64"}
```

For our use case it would make sense to load the correct Avro schema from a file, we are always going to be matching it by Event_${!json_field:event}_${!json_field:metadata.schema_version} and our Benthos docker image would have a folder `./avro/` with all known schema types for each event and each version as the schema evolves.

**Comment by Jeffail** [comment]
Hey @gregorskii, yeah when parsing and moving dates in Benthos with JSON they're always going to be numeric (float64) or string types.

There's possibly a setting for this within goavro, it makes sense to be able to convert a `float64` to `time.Time` when the schema specifies it. Otherwise we'll need to hack something together, maybe I can walk the schema and do it myself.

I can add dynamic file loading. From what I understand of avro it's fairly common to dynamically pull new schema versions, in which case we'll want to cache the various schemas with a TTL.

**Comment by torfjor** [comment]
We're currently consuming avro messages from kafka that are produced on the confluent platform with the schema registry. These messages have a [header](https://docs.confluent.io/platform/current/schema-registry/serdes-develop/index.html#wire-format) with a magic byte and a 4 byte ID.

Seeing as there's no schema registry support in benthos yet, we have to slice off the first five bytes of each message. It would be nice if benthos could parse the header and look up the schema if the config provided a path to a schema registry. Looks like there are some libraries that support schema registries already:

- [lensesio/schema-registry](https://github.com/lensesio/schema-registry)
- [riferrei/srclient](https://github.com/riferrei/srclient)

**Comment by SumuduSP** [comment]
> We're currently consuming avro messages from kafka that are produced on the confluent platform with the schema registry. These messages have a [header](https://docs.confluent.io/platform/current/schema-registry/serdes-develop/index.html#wire-format) with a magic byte and a 4 byte ID.
> 
> Seeing as there's no schema registry support in benthos yet, we have to slice off the first five bytes of each message. It would be nice if benthos could parse the header and look up the schema if the config provided a path to a schema registry. Looks like there are some libraries that support schema registries already:
> 
> * [lensesio/schema-registry](https://github.com/lensesio/schema-registry)
> * [riferrei/srclient](https://github.com/riferrei/srclient)

Hey @torfjor  Can I know whether you used benthos processors while consuming avro messages from kafka?

**Comment by torfjor** [comment]
Hi. Yes, we have a couple of instances of it running in production.

**Comment by SumuduSP** [comment]
> Hi @SumuduSP. Yes, we have a couple of instances of it running in production.
OK, @torfjor I am trying to use avro processor as mentioned in this thread to convert avro messages from kafka to_json but not able to do it with as I am getting this error "failed to parse schema: cannot unmarshal schema JSON: unexpected end of JSON input". Please let me know if you aware of this issue or If you know any other approach for this?

**Comment by Jeffail** [comment]
Quick update, I've added an experimental `schema_registry_decode` processor: https://www.benthos.dev/docs/components/processors/schema_registry_decode which makes decoding much easier. Currently only avro is supported but there's not much holding back the other formats.

I'm keeping this issue open for the purpose of adding a `schema_registry_encode` processor. It needs some thought as to what capabilities we expect this one to have as it's not as passive as the decode operation.

**Comment by hnorthcott** [comment]
Utilizing the `schema_registry_decode` processor now which works great. Unfortunately I cannot use the  `schema_registry_decode` processor if the Kafka topic is not encoded with the schema id and only in binary avro. My intended flow would be: 

Benthos job 1: Kafka Topic 1 -> Kafka Topic 2
Benthos job2: Kafka Topic 2 -> X
Python/Java/Go: Kafka Topic 2 -> X

This affects other AvroConsumers outside of Benthos consuming from Topic 2 that I produce to with Benthos. Since it is not encoded correctly in the specific [wire format](https://docs.confluent.io/platform/current/schema-registry/serdes-develop/index.html#supported-formats) that the schema registry follows. 

`schema_registry_encode` processor would be a huge help.

**Comment by hnorthcott** [comment]
Hi @Jeffail! Thanks so much for working on this. I have tested and integrated the schema registry encode into my existing Benthos yamls, everything is working as expected. I did run into one issue that I did find a solution for but don't know if you know a better workaround/fix for the actual plugin. 

If my avro schema has fields that the type is a union which is very common for example: 
` avsc
{
            "name": "key",
            "type": ["null", "string"]
},
`
Where the field can either be null or string, but defaults to null. The output root message needs to be uniquely formatted for these fields or else an error occurs.

The output message needs to be in golang map string interface format to successfully encode properly. If it is not formatted, I get the below error and no message pushed to the output Kafka topic. 
` schema.namespace.test" field "key": value does not match its schema: cannot encode binary union: non-nil Union values ought to be specified with Go map[string]interface{} `

This is the bloblang I currently have to do in order for it to encode properly. 
` 
pipeline:
    processors:
        - bloblang: |
              root = this
              let pre_map =  "{\"string\":"
              let close_map = "\"}"
              root.key = $pre_map + "\"" + this.key  + $close_map
       - bloblang: |
              root = this
              root.key =  this.key.parse_json()
`

Outside of this minor issue, it is working as expected for all non-union fields! Thanks so much for looking into this further, let me know if I can provide any other details.
